src/USER-DRUDE/pair_coul_tt.cpp

```diff
@@ -0,0 +1,446 @@
+
+void PairCoulTT::allocate()
+{
+  allocated = 1;
+  int n = atom->ntypes;
+
+  memory->create(setflag,n+1,n+1,"pair:setflag");
+  for (int i = 1; i <= n; i++)
+    for (int j = i; j <= n; j++)
+      setflag[i][j] = 0;
+
+  memory->create(cutsq,n+1,n+1,"pair:cutsq");
+
+  memory->create(cut,n+1,n+1,"pair:cut");
+  memory->create(scale,n+1,n+1,"pair:scale");
+  memory->create(b, n + 1, n + 1, "pair:b");
+  memory->create(c, n + 1, n + 1, "pair:c");
+  memory->create(ntt, n + 1, n + 1, "pair:ntt");
+}
+
+/* ----------------------------------------------------------------------
+   global settings
+------------------------------------------------------------------------- */
+
+void PairCoulTT::settings(int narg, char **arg)
+{
+  if (narg != 2) error->all(FLERR, "Illegal pair_style command");
+
+  n_global = utils::inumeric(FLERR,arg[0],false,lmp);
+  cut_global = utils::numeric(FLERR,arg[1],false,lmp);
+
+  // reset cutoffs that have been explicitly set
+
+  if (allocated) {
+    int i,j;
+    for (i = 1; i <= atom->ntypes; i++)
+      for (j = i; j <= atom->ntypes; j++)
+        if (setflag[i][j]) {
+            ntt[i][j] = n_global;
+          cut[i][j] = cut_global;
+        }
+  }
+}
+
+/* ----------------------------------------------------------------------
+   set coeffs for one or more type pairs
+------------------------------------------------------------------------- */
+
+void PairCoulTT::coeff(int narg, char **arg)
+{
+  if (narg < 3 || narg > 6)
+    error->all(FLERR,"Incorrect args for pair coefficients");
+  if (!allocated) allocate();
+
+  int ilo,ihi,jlo,jhi;
+
+  utils::bounds(FLERR,arg[0],1,atom->ntypes,ilo,ihi,error);
+  utils::bounds(FLERR,arg[1],1,atom->ntypes,jlo,jhi,error);
+
+  double b_one = utils::numeric(FLERR,arg[2],false,lmp);
+  double c_one = utils::numeric(FLERR,arg[3],false,lmp);
+  int n_one = n_global;
+  double cut_one = cut_global;
+  if (narg >= 5) n_one = utils::inumeric(FLERR,arg[4],false,lmp);
+  if (narg == 6) cut_one = utils::numeric(FLERR,arg[5],false,lmp);
+
+  if (n_one > n_global)
+    error->all(FLERR, "Incorrect coefficients for pair style coul/tt: n should not be larger than global setting");
+
+  int count = 0;
+  for (int i = ilo; i <= ihi; i++) {
+    for (int j = MAX(jlo,i); j <= jhi; j++) {
+      b[i][j] = b_one;
+      c[i][j] = c_one;
+      ntt[i][j] = n_one;
+      cut[i][j] = cut_one;
+      scale[i][j] = 1.0;
+      setflag[i][j] = 1;
+      count++;
+    }
+  }
+
+  if (count == 0) error->all(FLERR,"Incorrect args for pair coefficients");
+}
+
+
+/* ----------------------------------------------------------------------
+   init specific to this pair style
+------------------------------------------------------------------------- */
+
+void PairCoulTT::init_style()
+{
+  if (!atom->q_flag)
+    error->all(FLERR,"Pair style coul/tt requires atom attribute q");
+  int ifix;
+  for (ifix = 0; ifix < modify->nfix; ifix++)
+    if (strcmp(modify->fix[ifix]->style,"drude") == 0) break;
```

Comment:
please use utils::strmatch() here. This can be particularly useful in case somebody writes accelerated versions of the styles (since you would match against "^drude", which is ignoring suffixes).

Reply:
I have replaced `strcmp` by `utils::strmatch`. I forgot to change the regular expressions. I will add, recompile, rerun examples, and let you know.

Reply:
There are 4 entries of `strcmp`:
line 289: `   if (strcmp(modify->fix[ifix]->style,"drude") == 0) break;`
line 441-444:
```
  if (strcmp(str,"scale") == 0) return (void *) scale;
  if (strcmp(str,"b") == 0) return (void *) b;
  if (strcmp(str,"c") == 0) return (void *) c;
  if (strcmp(str,"ntt") == 0) return (void *) ntt;
```
If I replace all `strcmp` by `utils::strmatch`, I get the next error
`ERROR: Coulomb styles of pair hybrid sub-styles do not match (../pair_hybrid.cpp:1024)`

If I replace only the line 289, I get segmentation fault:

```
[Feynman:86022] *** Process received signal ***
[Feynman:86022] Signal: Segmentation fault: 11 (11)
[Feynman:86022] Signal code: Address not mapped (1)
[Feynman:86022] Failing at address: 0x34
[Feynman:86022] [ 0] 0   libsystem_platform.dylib            0x00007fff7406942d _sigtramp + 29
[Feynman:86022] [ 1] 0   ???                                 0x0000000000000000 0x0 + 0
[Feynman:86022] [ 2] 0   lmp_macos_mpi                       0x000000010cc5a06c _ZN9LAMMPS_NS10PairHybrid7computeEii + 620
[Feynman:86022] [ 3] 0   lmp_macos_mpi                       0x000000010cead63a _ZN9LAMMPS_NS6Verlet5setupEi + 474
[Feynman:86022] [ 4] 0   lmp_macos_mpi                       0x000000010ce426bd _ZN9LAMMPS_NS3Run7commandEiPPc + 2317
[Feynman:86022] [ 5] 0   lmp_macos_mpi                       0x000000010ca75fc4 _ZN9LAMMPS_NS5Input15command_creatorINS_3RunEEEvPNS_6LAMMPSEiPPc + 52
[Feynman:86022] [ 6] 0   lmp_macos_mpi                       0x000000010ca77fd1 _ZN9LAMMPS_NS5Input15execute_commandEv + 3521
[Feynman:86022] [ 7] 0   lmp_macos_mpi                       0x000000010ca76c3a _ZN9LAMMPS_NS5Input4fileEv + 810
[Feynman:86022] [ 8] 0   lmp_macos_mpi                       0x000000010c6598cb main + 75
[Feynman:86022] [ 9] 0   libdyld.dylib                       0x00007fff73e707fd start + 1
[Feynman:86022] [10] 0   ???                                 0x0000000000000003 0x0 + 3
[Feynman:86022] *** End of error message ***
Segmentation fault: 11
```
If I keep `strcmp` but use `"^drude"` instead of `"drude"`, I get
`ERROR: Pair coul/tt requires fix drude (../pair_coul_tt.cpp:290)`

If I do not change anything in `pair_coul_tt.cpp` keeping `strcmp` and old pattern for comparison, everything works fine.
(it seems that I forgot to `make pu` when I was testing `utils::strmatch` last time)

I'm compiling with `mpic++  -O3 -std=c++11`.

Could you please take a look at my `pair_coul_tt.cpp` file? Thank you in advance.

Reply:
you have the semantics of `utils::strmatch()` wrong hence the issues. The strmatch() function is most useful, if you want a partial match, e.g. for variants with different suffixes. In case of an exact match, like in the extract function, the original strcmp() will do as well, for strmatch() you would have to match against "^name$" because without it will behave like strstr() since it matches a regular expression substring

Reply:
Thank you a lot for your explanation!